alacritty/src/display/mod.rs

```diff
@@ -374,6 +378,77 @@ pub struct Display {
     meter: Meter,
 }
 
+/// The state of the input method on the `Display`.
+#[derive(Debug, Default)]
+pub struct Ime {
+    /// Whether the IME is enabled on the given window.
+    enabled: bool,
+
+    /// Current IME preedit.
+    preedit: Option<Preedit>,
+}
+
+impl Ime {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    #[inline]
+    pub fn set_enabled(&mut self, is_enabled: bool) {
+        if is_enabled {
+            self.enabled = is_enabled
+        } else {
+            // Clear state when disabling IME.
+            *self = Default::default();
```

Comment:
Well, we ensure that we don't have preedit. That's the point. We use preedit to determine whether we should show cursor or not. In general it can be like that, if the server sends you disable right away or you change the focus of the window.